From 4bbdb3b88c5f833dea35d6e611e66e43772f21e4 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Tue, 1 Sep 2015 14:23:04 +0000
Subject: [PATCH] [CONFIGURATION-609] Use the correct file system for include
 files.

The FileHandler used for loading include properties dropped some of the
properties configured for the builder. This has been fixed.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1700564 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfiguration.java              |  1 +
 .../TestPropertiesConfiguration.java          | 53 ++++++++++++++++++-
 2 files changed, 53 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
index ae5f115dcf..5747b697bb 100644
--- a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
@@ -1417,6 +1417,7 @@ private void loadIncludeFile(String fileName) throws ConfigurationException
         }
 
         FileHandler fh = new FileHandler(this);
+        fh.setFileLocator(locator);
         fh.load(url);
     }
 
diff --git a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
index f68a5aa967..459080da8d 100644
--- a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
@@ -27,15 +27,18 @@
 import static org.junit.Assert.fail;
 
 import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.FileReader;
 import java.io.FileWriter;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.Reader;
 import java.io.StringReader;
 import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
 import java.io.Writer;
 import java.net.HttpURLConnection;
 import java.net.URL;
@@ -50,14 +53,18 @@
 
 import org.apache.commons.configuration2.SynchronizerTestImpl.Methods;
 import org.apache.commons.configuration2.builder.FileBasedBuilderParametersImpl;
+import org.apache.commons.configuration2.builder.FileBasedConfigurationBuilder;
 import org.apache.commons.configuration2.builder.combined.CombinedConfigurationBuilder;
+import org.apache.commons.configuration2.builder.fluent.Parameters;
 import org.apache.commons.configuration2.convert.DefaultListDelimiterHandler;
 import org.apache.commons.configuration2.convert.DisabledListDelimiterHandler;
 import org.apache.commons.configuration2.convert.LegacyListDelimiterHandler;
 import org.apache.commons.configuration2.convert.ListDelimiterHandler;
 import org.apache.commons.configuration2.event.ConfigurationEvent;
 import org.apache.commons.configuration2.ex.ConfigurationException;
+import org.apache.commons.configuration2.io.DefaultFileSystem;
 import org.apache.commons.configuration2.io.FileHandler;
+import org.apache.commons.configuration2.io.FileSystem;
 import org.apache.commons.lang3.mutable.MutableObject;
 import org.junit.Before;
 import org.junit.Test;
@@ -143,7 +150,7 @@ public void testAppend() throws Exception
 
     /**
      * Checks for a property without a value.
-     * 
+     *
      * @param key the key to be checked
      */
     private void checkEmpty(String key)
@@ -1122,6 +1129,50 @@ public void testEscapeQuote() throws ConfigurationException
         assertEquals("Wrong value", text, c2.getString(PROP_NAME));
     }
 
+    /**
+     * Tests whether the correct file system is used when loading an include
+     * file. This test is related to CONFIGURATION-609.
+     */
+    @Test
+    public void testLoadIncludeFileViaFileSystem() throws ConfigurationException
+    {
+        conf.clear();
+        conf.addProperty("include", "include.properties");
+        saveTestConfig();
+
+        FileSystem fs = new DefaultFileSystem()
+        {
+            @Override
+            public InputStream getInputStream(URL url)
+                    throws ConfigurationException
+            {
+                if (url.toString().endsWith("include.properties"))
+                {
+                    try
+                    {
+                        return new ByteArrayInputStream(
+                                "test.outcome = success".getBytes("UTF-8"));
+                    }
+                    catch (UnsupportedEncodingException e)
+                    {
+                        throw new ConfigurationException("Unsupported encoding",
+                                e);
+                    }
+                }
+                return super.getInputStream(url);
+            }
+        };
+        Parameters params = new Parameters();
+        FileBasedConfigurationBuilder<PropertiesConfiguration> builder =
+                new FileBasedConfigurationBuilder<PropertiesConfiguration>(
+                        PropertiesConfiguration.class);
+        builder.configure(params.fileBased().setFile(testSavePropertiesFile)
+                .setBasePath(ConfigurationAssert.OUT_DIR.toURI().toString())
+                .setFileSystem(fs));
+        PropertiesConfiguration configuration = builder.getConfiguration();
+        assertEquals("success", configuration.getString("test.outcome"));
+    }
+
     /**
      * Helper method for testing the content of a list with elements that
      * contain backslashes.
